Unosquare.CodeChallenge/WarehouseServices/Services/CompaniesServices.cs

```diff
@@ -8,50 +8,91 @@
 using WarehouseServices.Contractor;
 using WarehouseCoreLib.Base;
 using WarehouseRepositories.Repositories;
+using Microsoft.EntityFrameworkCore;
 
 namespace WarehouseServices.Services
 {
     public class CompaniesServices : Service<Company>, IWarehouseService<CompaniesServices>
     {
-        private CompanyRepository repository;
+        private WarehouseDbContext dbcontext;
 
         public CompaniesServices(WarehouseDbContext context)
         {
-            repository = new CompanyRepository();
-            repository.dbcontext = context;
+            dbcontext = context;
         }
 
-        public IEnumerable<Company> getAllCompanies()
+        public IEnumerable<Company> GetAllCompanies(int limit = 0)
         {
-            return repository.findAll();
+
+            if (limit == 0) return dbcontext.Companies;
+            return dbcontext.Companies.Take(limit);
+        }
+
+        public async Task<List<Company>> GetAllCompaniesAsync(int limit = 0)
+        {
+            return await dbcontext.Companies.ToListAsync();
+        }
+
+        public Company GetCompany(int id)
+        {
+            return dbcontext.Companies.Find(id);
+        }
+
+        public async Task<Company> GetCompanyAsync(int id)
+        {
+            return  await dbcontext.Companies.FindAsync(id);
+        }
+
+        public void AddCompany(Company company)
+        {
+            dbcontext.Companies.Add(company);
+            dbcontext.SaveChanges();
+        }
+
+        public async Task AddCompanyAsync(Company company)
+        {
+            await dbcontext.Companies.AddAsync(company);
+            await dbcontext.SaveChangesAsync();
+        }
+
+        public void DeleteCompany(Company company)
+        {
+            dbcontext.Companies.Remove(company);
+            dbcontext.SaveChanges();
         }
 
-        public Company getCompany(int id)
+        public async Task DeleteCompanyAsync(Company company)
         {
-            return repository.getById(id);
+            await Task.Run(() =>{
```

Comment:
Why use Task Run?

Reply:
Since there are not RemoveAsync, I create a task to run it

Reply:
I made the refactor  on the commit f4f6e2d1a270fce4c21d7acf8db531a02cde94fe, removing the task